### ISSUE TITLE
Possible typo in RegExp in splitNumbers

### ISSUE THREAD

**Issue by lrhn** [josdejong/mathjs#2363]
Just saw the RegExp `/^0*?(-?)(\d+\.?\d*)(e([+-]?\d+))?$/` in [utils/numbers.js](https://github.com/josdejong/mathjs/blob/4ce71856afbcf1a0aaad30b70afc99c7606173c0/src/utils/number.js#L311).

I think the `0*?` should be after the `(-?)`.

Consider something like `/^(-?)0*?((?:0|[1-9]\d*)(?:\.\d*)?)(e([+-]?\d+))?$/`.
(It also optimizes for avoiding too much backtracking).

**Comment by josdejong** [comment]
Thanks, I think you're right. I'm just wondering now what the `0*?` was for in the first place. When I remove it, none of the unit tests break. So maybe it's just fine to remove that part.

**Comment by lrhn** [comment]
I'm guessing the `0*` is to allow extra leading zeros in the input, like `"0001.2e4"`. Just being permissive, not something that's needed if the input comes from `Number.toString()` or similar normalized strings.

**Comment by josdejong** [comment]
yeah that makes sense. In that case I can just remove this first part `0*?` and turn the regex from:

```
/^0*?(-?)(\d+\.?\d*)(e([+-]?\d+))?$/
```

into:

```js
/^(-?)(\d+\.?\d*)(e([+-]?\d+))?$/
```

I see you propose more changes, do these offer additional improvements?

**Comment by lrhn** [comment]
The other changes are only potential performance improvements. If it's not a performance critical regexp, it probably doesn't matter. Even then, it's likely mainly theoretical.

The RegExp part `0*?(-?)(\d+\.?\d*)` can match a `0` by both `0` and the two `\d`s because the `-` and `.` are both optional.
That makes it possible to match `000000...000000000` (n zeros) in roughly O(n^3) ways. That doesn't matter if the RegExp *matches* the input because it'll find *some* match fairly quickly, but if the text doesn't match, like the string `"00000000000000000000000000e"`, it will potentially go through all the possible ways to split the zeros into three parts before rejecting the string.

Failing to match *in this way* is probably going to be rare. Removing the leading `0*` will also reduce the complexity to O(n^2).
All in all, probably not a problem in practice.

**Comment by josdejong** [comment]
Thanks for your clarification. I tested a bit more and the `0*?` in the current regex is indeed just redundant: it should be after the `(-?)`, but there it is redundant with the `\d+` that follows.

I removed the `0*?` part now in https://github.com/josdejong/mathjs/commit/e8a50259e7a97bcfeb01dc241cf2d06b7d2f730c.

I prefer to keep the regex as short and simple as possible for now, and only change it when there are measurable performance improvements. I guess that the optimized regex you propose is faster, but it's also quite a bit longer and hence harder to read/understand. It's a trade-off between simple/readable code and optimization :). Thanks for all you thinking along!
